dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/HostStatusTracker.java

```diff
@@ -128,9 +128,45 @@ public boolean inactiveSetChanged(Collection<Host> hostsUp, Collection<Host> hos
 	 * @return true/false indicating whether the set of hosts has changed or not.
 	 */
 	public boolean checkIfChanged(Collection<Host> hostsUp, Collection<Host> hostsDown) {
-		return activeSetChanged(hostsUp) || inactiveSetChanged(hostsUp, hostsDown);
+		boolean changed = activeSetChanged(hostsUp) || inactiveSetChanged(hostsUp, hostsDown);
+		if (changed) logChangedHosts(hostsUp, hostsDown);
+		return changed;
 	}
-	
+
+	/**
+	 * Helper method that checks if anything has changed b/w the current state and the new set of hosts up and down and
+	 * logs the difference in hosts.
+	 * @param hostsUp
+	 * @param hostsDown
+	 */
+	public void logChangedHosts(Collection<Host> hostsUp, Collection<Host> hostsDown) {
+		if (!logger.isDebugEnabled()) return;
+
+		Set<Host> changedHostsUp = new HashSet<>();
+		changedHostsUp.addAll(hostsUp);
+		for (Host host : activeHosts) {
+			if (changedHostsUp.contains(host)) {
+				changedHostsUp.remove(host);
+			}
+		}
+
+		Set<Host> changedHostsDown = new HashSet<>();
+		changedHostsDown.addAll(hostsDown);
+		for (Host host : inactiveHosts) {
+			if (changedHostsDown.contains(host)) {
+				changedHostsDown.remove(host);
+			}
+		}
+
+		for (Host host : changedHostsUp) {
+			logger.debug("New host up: " + host.getHostAddress());
```

Comment:
log4j best practices:

- Log4j logging can be efficient (lazy argument evaluation) when parameters can be printed with {} instead of concatenating the string representation of object. The above can be changed as - 

`if (logger.isDebugEnabled()) {
  logger.debug("New host up: {}", host.getHostAddress());
}
`

- It is also good practice to use isDebugEnabled() for debug logs as it can avoid expensive computation of log messages in the log4j stack.

Reply:
Done, thanks!